cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ZookeeperSetOperator.java

```diff
@@ -54,4 +102,225 @@ public static boolean needsRollingUpdate(StatefulSetDiff diff) {
+            kmf.init(setupKeyStore(clusterCaKeySecret, x509, password, caCertCO, coCertKey), password);
+
+            TrustManagerFactory tmf = TrustManagerFactory.getInstance("X509");
+            tmf.init(setupTrustStore(password, caCertCO));
+            SSLContext ctx = SSLContext.getInstance("TLS");
+            ctx.init(kmf.getKeyManagers(), tmf.getTrustManagers(), null);
+            factory = ctx.getSocketFactory();
+        } catch (CertificateException | NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException | KeyManagementException e) {
+            log.error(e.getMessage());
+        }
+        return factory;
+    }
+
+    private boolean isLeader(Pod pod, SSLSocketFactory factory) {
+        int port = 2181;
+        boolean leader = false;
+        try {
+            String host = pod.getStatus().getPodIP();
+
+            SSLSocket socket = null;
+            Socket plainSocket = factory.createSocket();
+            if (plainSocket instanceof SSLSocket) {
+                // findbugs plugin does not like direct overtyping
+                socket = (SSLSocket) plainSocket;
+            }
+            log.debug("Connecting client to {}:{}", host, port);
+            try {
+                socket.connect(new InetSocketAddress(host, port), 10_000);
+            } catch (ConnectException | SocketTimeoutException e) {
+                log.error("Could not connect " + e.getMessage());
+            }
+            try {
+                log.debug("Starting handshake with {}", socket.getRemoteSocketAddress());
+                try {
+                    socket.startHandshake();
+                    PrintWriter out = new PrintWriter(
+                            new BufferedWriter(
+                                    new OutputStreamWriter(
+                                            socket.getOutputStream(), StandardCharsets.UTF_8)));
+
+                    out.println("stat");
+                    out.flush();
+
+                    BufferedReader in = new BufferedReader(
+                            new InputStreamReader(
+                                    socket.getInputStream(), StandardCharsets.UTF_8));
+
+                    String inputLine;
+                    while ((inputLine = in.readLine()) != null) {
+                        log.debug(inputLine);
+                        if (inputLine.equals("Mode: leader")) {
+                            leader = true;
+                        }
+                    }
+                    in.close();
+                    out.close();
+                } catch (SSLHandshakeException e) {
+                    log.error("handshake exception - " + e.getMessage());
+                    e.printStackTrace();
+                } catch (SocketException w) {
+
+                }
+            } finally {
+                socket.close();
+            }
+        } catch (NullPointerException | IOException e) {
```

Comment:
Catching NPE is generally a really bad idea. You don't know where that NPE originates. At the time you write it you might think you know which line of code is throwing the NPE. But remember that any of the code in the `try` block could be throwing it, in circumstances you have decided are OK, but also in places and ways that you have no idea of right now. It might be OK today, but over time this will get refactored and someone else might end up calling some code in the try block with a null by accident. Catching NPE makes these things much much harder to diagnose and fix. Worse you're not even logging the exception (with stacktrace) in the `catch`, so all NPEs thrown in the `try` will look identical in the logs. Remember: It could be you who ends up trying to fix a problem with only the logs to try to reproduce it. Do yourself a favour and include more information rather than less.

Anyway, getting back to catching NPE: If it's your code which is throwing the NPE you should be checking for null _before_ it throws. If it's some 3rd party code which is throwing it then either you're calling it wrong (i.e. passing some reference which – possibly transitively in the object graph – is null), or that 3rd party code has a bug.